### ISSUE TITLE
[no-misused-promises] assignment async function to non-async function

### ISSUE THREAD

**Issue by jet2jet** [typescript-eslint/typescript-eslint#3728]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-misused-promises": ["error"]
  }
}
```

```TSX
interface Props {
    cb: () => void;
}
function foo(cb: () => void) { }
function bar(props: Props) { }

// 1. will be reported
foo(async () => { });
// 2. will not be reported
bar({ cb: async () => { } });

const fn1 = async () => { };
// 3. will be reported
foo(fn1);

// 4. will not be reported
const fn2: () => void = async () => { };
foo(fn2);

// 5. will not be reported
const elem = <button onClick={fn1}></button>
```

tsconfig: 

```JSON
{
  "compilerOptions": {
    "target": "ESNEXT",
    "jsx": "preserve",
    "strict": true
  }
}
```

**Expected Result**

All 1 to 5 will be reported as an error

**Actual Result**

Only 1 and 3 will be reported

**Additional Info**

I wonder this is not actually a bug but a 'not supported' case, but I'm happy if such the case is also supported with the option `checksVoidReturn`.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.29.1` |
| `@typescript-eslint/parser`        | `4.29.1` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `12.16.2` |

**Comment by thw0rted** [comment]
I think examples 2 and 4 are a TS problem, not something the linter can check.  It's to do with the way the typechecker handles covariance, though every time it comes up I have to go read some definitions to make sure I'm getting the terms right.  It's nothing to do with promises, you can also write

````ts
function fn1() { return "Hello"; }

const fn2: () => void = fn1;

const x: {a: () => void} = {a: fn1};
````

without anything getting flagged.  You told the typechecker `fn2()` returns `void`, but it doesn't; same with `x.a()`.

FYI, while performing the search that led me to this issue, I found https://github.com/typescript-eslint/typescript-eslint/issues/1744, which looks related.  Probably worth trakcing.

**Comment by jet2jet** [comment]
@thw0rted thanks to the comment.
I'll close this issue as a duplicate of #1744.
